internal/content_coding.go

```diff
@@ -370,19 +423,50 @@ func (d *ZlibDecoder) Decode(data []byte, maxDecompressionSize int64) ([]byte, e
 	return d.buf.Bytes(), nil
 }
 
+type ZstdDecoder struct {
+	decoder *zstd.Decoder
+}
+
+func NewZstdDecoder(options ...DecodingOption) (*ZstdDecoder, error) {
+	cfg := decoderConfig{maxDecompressionSize: DefaultMaxDecompressionSize}
+	for _, o := range options {
+		o(&cfg)
+	}
+
+	d, err := zstd.NewReader(nil, zstd.WithDecoderConcurrency(0), zstd.WithDecoderMaxWindow(uint64(cfg.maxDecompressionSize)))
+	return &ZstdDecoder{
+		decoder: d,
+	}, err
+}
+
+func (*ZstdDecoder) SetEncoding(string) {}
+
+func (d *ZstdDecoder) Decode(data []byte) ([]byte, error) {
+	return d.decoder.DecodeAll(data, nil)
+}
+
 // IdentityDecoder is a null decoder that returns the input.
-type IdentityDecoder struct{}
+type IdentityDecoder struct {
+	maxDecompressionSize int64
+}
 
-func NewIdentityDecoder() *IdentityDecoder {
-	return &IdentityDecoder{}
+func NewIdentityDecoder(options ...DecodingOption) *IdentityDecoder {
+	cfg := decoderConfig{maxDecompressionSize: DefaultMaxDecompressionSize}
+	for _, o := range options {
+		o(&cfg)
+	}
+
+	return &IdentityDecoder{
+		maxDecompressionSize: cfg.maxDecompressionSize,
+	}
 }
 
 func (*IdentityDecoder) SetEncoding(string) {}
 
-func (*IdentityDecoder) Decode(data []byte, maxDecompressionSize int64) ([]byte, error) {
+func (d *IdentityDecoder) Decode(data []byte) ([]byte, error) {
 	size := int64(len(data))
-	if size > maxDecompressionSize {
-		return nil, fmt.Errorf("size of decoded data: %d exceeds allowed size %d", size, maxDecompressionSize)
+	if size > d.maxDecompressionSize {
+		return nil, fmt.Errorf("size of decoded data: %d exceeds allowed size %d", size, d.maxDecompressionSize)
 	}
 	return data, nil
 }
```

Comment:
Please do not artificially constrain the `IdentityDecoder`! Just leave that one alone and ignore all options silently.